src/iceberg/update/update_partition_spec.h

```diff
@@ -0,0 +1,241 @@
+
+  /// \brief Hash function for TransformKey.
+  struct TransformKeyHash {
+    size_t operator()(const TransformKey& key) const {
+      return std::hash<int32_t>{}(key.first) ^
+             (std::hash<std::string>{}(key.second) << 1);
+    }
+  };
+
+  /// \brief Assign a new partition field ID.
+  int32_t AssignFieldId();
+
+  /// \brief Recycle or create a partition field.
+  ///
+  /// In V2, searches for a similar partition field in historical specs.
+  /// If not found or in V1, creates a new PartitionField.
+  PartitionField RecycleOrCreatePartitionField(int32_t source_id,
+                                               std::shared_ptr<Transform> transform,
+                                               const std::string* name);
+
+  /// \brief Internal implementation of AddField with resolved source ID and transform.
+  UpdatePartitionSpec& AddFieldInternal(const std::string* name, int32_t source_id,
+                                        std::shared_ptr<Transform> transform);
+
+  /// \brief Generate a partition field name from the source and transform.
+  std::string GeneratePartitionName(int32_t source_id,
+                                    const std::shared_ptr<Transform>& transform) const;
+
+  /// \brief Check if a transform is a time-based transform.
+  static bool IsTimeTransform(const std::shared_ptr<Transform>& transform);
+
+  /// \brief Check if a partition field uses void transform.
+  static bool IsVoidTransform(const PartitionField& field);
+
+  /// \brief Check for redundant time-based partition fields.
+  void CheckForRedundantAddedPartitions(const PartitionField& field);
+
+  /// \brief Handle rewriting a delete-and-add operation for the same field.
+  UpdatePartitionSpec& RewriteDeleteAndAddField(const PartitionField& existing,
+                                                const std::string* name);
+
+  /// \brief Internal helper to remove a field by transform key.
+  UpdatePartitionSpec& RemoveFieldByTransform(const TransformKey& key,
+                                              const std::string& term_str);
+
+  /// \brief Index the spec fields by name.
+  static std::unordered_map<std::string, PartitionField> IndexSpecByName(
+      const PartitionSpec& spec);
+
+  /// \brief Index the spec fields by (source_id, transform) pair.
+  static std::unordered_map<TransformKey, PartitionField, TransformKeyHash>
+  IndexSpecByTransform(const PartitionSpec& spec);
+
+  TableIdentifier identifier_;
+  std::shared_ptr<Catalog> catalog_;
+  std::shared_ptr<TableMetadata> base_metadata_;
+
+  // Configuration
+  int32_t format_version_;
+  std::shared_ptr<PartitionSpec> spec_;
+  std::shared_ptr<Schema> schema_;
+  bool case_sensitive_{true};
+  bool set_as_default_{true};
+  int32_t last_assigned_partition_id_;
+
+  // Indexes for existing fields
+  std::unordered_map<std::string, PartitionField> name_to_field_;
+  std::unordered_map<TransformKey, PartitionField, TransformKeyHash> transform_to_field_;
```

Comment:
```suggestion
  std::unordered_map<std::string, const PartitionField*> name_to_field_;
  std::unordered_map<TransformKey, const PartitionField*, TransformKeyHash> transform_to_field_;
```

It is unnecessary to use a copy here.